File: ghidra_scripts/decompile_fill.py
# @category TDS-Clone
# @runtime Jython
# decompile_fill.py - decompile cac ham lien quan FILL/ASK/SPREAD de tim nguon fill-ask.
# Pure ASCII (Jython 2.7).

print("=== decompile_fill.py START ===")
from ghidra.app.decompiler import DecompInterface
from ghidra.util.task import ConsoleTaskMonitor

OUT = r"C:\Users\Phuc\Desktop\tickdata\ghidra_scripts\decompile_fill_out.txt"

prog = currentProgram
base = prog.getImageBase().getOffset()
af = prog.getAddressFactory().getDefaultAddressSpace()
fm = prog.getFunctionManager()

def resolve(va):
    # thu ca VA (base 0x400000) va RVA (base 0)
    for cand in (va, va - 0x400000 + base, va + base):
        try:
            a = af.getAddress(cand)
            f = fm.getFunctionContaining(a)
            if f is not None:
                return a, f
        except:
            pass
    # neu khong co function, thu tao tai dia chi
    try:
        a = af.getAddress(va)
        return a, fm.getFunctionContaining(a)
    except:
        return None, None

di = DecompInterface()
di.openProgram(prog)
mon = ConsoleTaskMonitor()

# Cac dia chi (VA base 0x400000) da RE:
TARGETS = [
    (0x00F7E3F0, "TickProc_setBidAsk"),    # set model.bid/ask tu outTick
    (0x00F7A4FB, "OrderCheck_SLTP"),        # kiem tra SL/TP dung ask 0x60d0
    (0x00F5EEF0, "TesterTickLoop"),         # vong lap tick goi FetchNextTick+TickProc
    (0x00F84FB0, "FetchNextTick"),          # sinh tick (ask=bid+model.328)
    (0x00F83C00, "FxtLoader_validate"),     # doc header + spread
    (0x00F83640, "TestGen_spreadSetup"),    # "spread set to %d"
]

print("ImageBase = 0x%X" % base)
out = open(OUT, "w")
out.write("ImageBase=0x%X\n\n" % base)
for va, name in TARGETS:
    a, f = resolve(va)
    if f is None:
        out.write("### %s @ 0x%X : KHONG tim thay function\n\n" % (name, va))
        print("%s: no function" % name)
        continue
    out.write("### %s  (req 0x%X -> func %s @ %s)\n" % (name, va, f.getName(), f.getEntryPoint()))
    try:
        res = di.decompileFunction(f, 60, mon)
        if res and res.decompileCompleted():
            out.write(res.getDecompiledFunction().getC())
        else:
            out.write("  [decompile failed: %s]\n" % (res.getErrorMessage() if res else "null"))
    except:
        import traceback
        out.write("  [exception]\n" + traceback.format_exc())
    out.write("\n\n" + "="*70 + "\n\n")
    print("done %s" % name)
out.close()
print("=== WROTE %s ===" % OUT)

File: ghidra_scripts/DecompileFill.java
// Decompile cac ham FILL/ASK/SPREAD -> tim nguon fill-ask. Java (Ghidra 12).
// @category TDS-Clone
import ghidra.app.script.GhidraScript;
import ghidra.app.decompiler.DecompInterface;
import ghidra.app.decompiler.DecompileResults;
import ghidra.program.model.address.Address;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionManager;
import java.io.PrintWriter;

public class DecompileFill extends GhidraScript {
    public void run() throws Exception {
        String OUT = "C:\\Users\\Phuc\\Desktop\\tickdata\\ghidra_scripts\\decompile_fill_out.txt";
        long base = currentProgram.getImageBase().getOffset();
        DecompInterface di = new DecompInterface();
        di.openProgram(currentProgram);
        FunctionManager fm = currentProgram.getFunctionManager();

        long[] targets = {
            0x00F7E3F0L,  // TickProc: set model.bid/ask tu outTick
            0x00F7A4FBL,  // OrderCheck SL/TP dung ask 0x60d0
            0x00F7A050L,  // vung xu ly lenh khac
            0x00F5EEF0L,  // tester tick loop
            0x00F84FB0L,  // FetchNextTick
        };
        String[] names = {"TickProc","OrderCheckSLTP","OrderProc2","TesterTickLoop","FetchNextTick"};

        PrintWriter out = new PrintWriter(OUT);
        out.println("ImageBase=0x" + Long.toHexString(base));
        for (int i = 0; i < targets.length; i++) {
            Address a = toAddr(targets[i]);
            Function f = fm.getFunctionContaining(a);
            if (f == null) {
                // thu disassemble + tao function tai dia chi
                try { disassemble(a); } catch (Exception e) {}
                try { f = createFunction(a, names[i]); } catch (Exception e) {}
                if (f == null) f = fm.getFunctionContaining(a);
            }
            if (f == null) {
                out.println("### " + names[i] + " @ 0x" + Long.toHexString(targets[i]) + " : NO FUNCTION");
                out.println();
                continue;
            }
            out.println("### " + names[i] + "  req=0x" + Long.toHexString(targets[i])
                        + " func=" + f.getName() + " @ " + f.getEntryPoint());
            try {
                DecompileResults res = di.decompileFunction(f, 90, monitor);
                if (res != null && res.decompileCompleted())
                    out.println(res.getDecompiledFunction().getC());
                else
                    out.println("  [decompile failed: " + (res==null?"null":res.getErrorMessage()) + "]");
            } catch (Exception e) {
                out.println("  [exception] " + e);
            }
            out.println();
            out.println("======================================================================");
            out.println();
        }
        out.close();
        println("WROTE " + OUT);
    }
}
